src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs

```diff
@@ -274,7 +274,7 @@ private ReturnTypeChain BuildReturnTypes()
                     throw new InvalidOperationException($"Method {Operation.Name} has to have a return value");
                 }
 
-                if (responseType.TryCast<ModelTypeProvider>(out var modelType))
+                if (responseType is { IsFrameworkType: false, Implementation: ModelTypeProvider modelType })
```

Comment:
Is `IsFrameworkType: false` needed? If `Implementation` is set, should `IsFrameworkType` always be false?

Reply:
Implementation will throw exceptions when IsFrameworkType is false.
C# compiler compiles this into a sequence of if statements, therefore if we do not have IsFrameworkType: false, it throws exceptions.

Reply:
Something like this:
![image](https://github.com/Azure/autorest.csharp/assets/10554446/034929da-5edd-450d-9c3e-fe5beef1185a)

Reply:
Right. And why do we throw while accessing `Implemenation` instead of making it nullable?

Reply:
This CSharpType is a wrap and a union of `System.Type` and `TypeProvider`
It might benefit and make things easier to debug when things go wrong by throwing an exception instead of making it nullable - because they are so different.
And making it nullable leads to a result of we are using `!` everywhere or we have to check if it is not null, therefore it really does not validate anything.
In the situation now, if we are 100% sure, this cannot be a framework type, we could just use `type.Implementation` and do not need to add the nullable forgiving operator

Reply:
OK. This then ends up with checking `IsFrameworkType: false` everywhere we want to check the type of `Implementation`.

Reply:
And if we use `!` to bypass compiler check, it still throws at run time if it is null, which is basically the same as the throwing now.

Reply:
Yeah both ways have their shortage. But it is introduced a long time ago therefore it is not in the scope of changing it.
This is similar to the pattern when you have a union in typescript - you will have to assert it is the type before you could use any function or property on it (as far as I could remember)